### PR TITLE
Don't panic when wgpu swapchain frame is outdated

### LINKED ISSUE #799: Crashes when moving workspaces in bspwm
When I move my window to a new workspace in bspwm, it crashes when I switch to the workspace I moved it to. It's fine until I actually switch to that workspace.
It does not seem to happen on i3wm. I can resize the window without issue, only moving it between workspaces causes problems. I am on Arch Linux, using bbb4e4678f14b4b187f9537a32063440e727e919. I'm also using Picom as my compositor but I'm not sure if that matters, just mentioning it since the error talks about compositors.
<details>
    <summary><code>RUST_BACKTRACE=full cargo run --quiet</code></summary>
<pre><code>thread 'main' panicked at 'Next frame: Outdated', /home/user/.cargo/git/checkouts/iced-9e73d2fbe1fce35a/bbb4e46/wgpu/src/window/compositor.rs:126:52
stack backtrace:
   0:     0x558669fa8870 - std::backtrace_rs::backtrace::libunwind::trace::h25e12e0d899beba0
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/../../backtrace/src/backtrace/libunwind.rs:90:5
   1:     0x558669fa8870 - std::backtrace_rs::backtrace::trace_unsynchronized::h70e61195d6ae3df6
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x558669fa8870 - std::sys_common::backtrace::_print_fmt::hba93ab80d779695a
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x558669fa8870 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hf092b5883b4b2e50
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/sys_common/backtrace.rs:46:22
   4:     0x558669fca2dc - core::fmt::write::hf68bc350a8f2f0dc
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/fmt/mod.rs:1078:17
   5:     0x558669fa2bc2 - std::io::Write::write_fmt::hf66811b1bc767436
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/io/mod.rs:1517:15
   6:     0x558669faaa45 - std::sys_common::backtrace::_print::hd425a11bfe1f20f8
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/sys_common/backtrace.rs:49:5
   7:     0x558669faaa45 - std::sys_common::backtrace::print::h6d678795c1e61e13
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/sys_common/backtrace.rs:36:9
   8:     0x558669faaa45 - std::panicking::default_hook::{{closure}}::h78a02a4a0dee5e7e
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:208:50
   9:     0x558669faa59a - std::panicking::default_hook::h56eb7eda02f355a7
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:225:9
  10:     0x558669fab1e1 - std::panicking::rust_panic_with_hook::hb27ea14285131c61
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:591:17
  11:     0x558669faad27 - std::panicking::begin_panic_handler::{{closure}}::hc552fcee62aad17f
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:497:13
  12:     0x558669fa8d2c - std::sys_common::backtrace::__rust_end_short_backtrace::hb9f0aa9a78e885a0
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/sys_common/backtrace.rs:141:18
  13:     0x558669faac89 - rust_begin_unwind
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:493:5
  14:     0x558669fc7841 - core::panicking::panic_fmt::h12ac4570ea43d06f
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/panicking.rs:92:14
  15:     0x558669fc7663 - core::option::expect_none_failed::h096fa60f757b7204
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/option.rs:1268:5
  16:     0x5586688ba467 - core::result::Result<T,E>::expect::h890ed258bd7ab255
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:933:23
  17:     0x5586688e3299 - <iced_wgpu::window::compositor::Compositor as iced_graphics::window::compositor::Compositor>::draw::hf6e852b9c190702e
                               at /home/user/.cargo/git/checkouts/iced-9e73d2fbe1fce35a/bbb4e46/wgpu/src/window/compositor.rs:126:21
  18:     0x558668662e46 - iced_winit::application::run_instance::{{closure}}::h34f22718bde9423f
                               at /home/user/.cargo/git/checkouts/iced-9e73d2fbe1fce35a/bbb4e46/winit/src/application.rs:335:45
  19:     0x558668895cf9 - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::he2722b3e89b1a583
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  20:     0x558668665794 - iced_winit::application::run::{{closure}}::h2ddb137397ec7eaa
                               at /home/user/.cargo/git/checkouts/iced-9e73d2fbe1fce35a/bbb4e46/winit/src/application.rs:185:24
  21:     0x55866890c14d - winit::platform_impl::platform::sticky_exit_callback::hfe7dbe34d457a29f
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/mod.rs:736:5
  22:     0x558668749b92 - winit::platform_impl::platform::x11::EventLoop<T>::run_return::h89b5d02986ab9184
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/x11/mod.rs:320:21
  23:     0x55866874c434 - winit::platform_impl::platform::x11::EventLoop<T>::run::h664b113ea24d6e2d
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/x11/mod.rs:398:9
  24:     0x55866890be86 - winit::platform_impl::platform::EventLoop<T>::run::h8e6204e97a504661
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/platform_impl/linux/mod.rs:652:56
  25:     0x55866870f486 - winit::event_loop::EventLoop<T>::run::hda4bbddf02bf100c
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/winit-0.24.0/src/event_loop.rs:154:9
  26:     0x558668665008 - iced_winit::application::run::h967e43e0082754b0
                               at /home/user/.cargo/git/checkouts/iced-9e73d2fbe1fce35a/bbb4e46/winit/src/application.rs:160:5
  27:     0x558668935553 - iced::application::Application::run::hadd455378de7b6a6
                               at /home/user/.cargo/git/checkouts/iced-9e73d2fbe1fce35a/bbb4e46/src/application.rs:212:16
  28:     0x5586689358eb - easyworship_rs::main::{{closure}}::h821e4fd1f2412c5d
                               at /home/user/church/worship_automation/easyworship-rs/src/main.rs:18:5
  29:     0x55866889564c - <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll::h9bea4dc3d351203e
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/future/mod.rs:80:19
  30:     0x5586688de826 - tokio::park::thread::CachedParkThread::block_on::{{closure}}::h60aa4b0b75b25f40
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/park/thread.rs:263:54
  31:     0x558668921818 - tokio::coop::with_budget::{{closure}}::h5c173808ea8d93aa
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/coop.rs:106:9
  32:     0x55866887fcc3 - std::thread::local::LocalKey<T>::try_with::hdbf260bdde87b8f9
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:272:16
  33:     0x55866887f0e2 - std::thread::local::LocalKey<T>::with::hbb43c967ffef6c0b
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/local.rs:248:9
  34:     0x5586688de54d - tokio::coop::with_budget::h389b605b1ba351f2
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/coop.rs:99:5
  35:     0x5586688de54d - tokio::coop::budget::h843a22394c9a9772
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/coop.rs:76:5
  36:     0x5586688de54d - tokio::park::thread::CachedParkThread::block_on::h859bd3790640f566
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/park/thread.rs:263:31
  37:     0x558668757c88 - tokio::runtime::enter::Enter::block_on::h72d45b97b681d639
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/runtime/enter.rs:151:13
  38:     0x558668758f10 - tokio::runtime::thread_pool::ThreadPool::block_on::he928f8a152f41142
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/runtime/thread_pool/mod.rs:71:9
  39:     0x558668758d84 - tokio::runtime::Runtime::block_on::ha73f3db41ce51f92
                               at /home/user/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.4.0/src/runtime/mod.rs:452:43
  40:     0x5586688ea7db - easyworship_rs::main::hf8a0bf0aea6f54b2
                               at /home/user/church/worship_automation/easyworship-rs/src/main.rs:16:1
  41:     0x5586686f2e42 - core::ops::function::FnOnce::call_once::h68f1866a10bfe62c
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
  42:     0x55866888013a - std::sys_common::backtrace::__rust_begin_short_backtrace::h90289a3e7d1d2ab1
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:125:18
  43:     0x558668757a86 - std::rt::lang_start::{{closure}}::h51bff070e1c4b017
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:66:18
  44:     0x558669fab607 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h78040f802d89ccdc
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/core/src/ops/function.rs:259:13
  45:     0x558669fab607 - std::panicking::try::do_call::h6853cad536dd09a1
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:379:40
  46:     0x558669fab607 - std::panicking::try::h827495f03a9fbb9a
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panicking.rs:343:19
  47:     0x558669fab607 - std::panic::catch_unwind::h4bdf17571090eb17
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/panic.rs:396:14
  48:     0x558669fab607 - std::rt::lang_start_internal::h2f319c33bb013f29
                               at /rustc/cb75ad5db02783e8b0222fee363c5f63f7e2cf5b/library/std/src/rt.rs:51:25
  49:     0x558668757a57 - std::rt::lang_start::hc1cad30da869ecea
                               at /home/user/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/rt.rs:65:5
  50:     0x5586688ea90a - main
  51:     0x7f2d773d3b25 - __libc_start_main
  52:     0x5586686605be - _start
  53:                0x0 - <unknown>
[1]    125783 segmentation fault (core dumped)  RUST_BACKTRACE=full cargo run --quiet</pre></code>
</details>

### PR DESCRIPTION
**Pull request by BillyDM**
The compositor in `iced_wgpu` should not panic when the wgpu swapchain returns an outdated error, which can happen when the window is resized. Instead, simply try drawing again the next frame.

I haven't encountered this issue much with the `winit` version of iced, but it is happening consistently with my own `iced_baseview` backend. Could be differences in timing.

Fixes #799.

### CODE REVIEW

**Comment by BillyDM** [comment]
We figured out this issue is caused by the fact that X11 events are sent asynchronously. So there is no way to ensure that a resize event won't be sent during the time between `MainEventsCleared` and `RedrawRequested`. We suspect that winit's timing just makes it less likely for this race condition to occur.

We added a fix to `baseview` to make it less likely to happen as well, but we should still at least not panic if this situation occurs.

**Comment by PolyMeilex** [comment]
Happen to me on winit multiple times as well, so it would be nice to get this merged

**Comment by BillyDM** [comment]
It turns out, some NVidia drivers on Linux will also panic like this right when the applications starts. This only appears to happen when the window is a parented window, such as audio plugins made with `iced_baseview`.

File: examples/integration/src/main.rs [hunk 1]
```diff
@@ -168,55 +168,64 @@ pub fn main() {
...
-
-                local_pool.run_until_stalled();
+                        local_pool.run_until_stalled();
+                    }
+                    Err(error) => match error {
+                        wgpu::SwapChainError::Outdated => {
+                            // Try rendering again next frame.
+                            window.request_redraw();
+                        }
+                        _ => panic!("Swapchain error: {:?}", error),
```
Comment (derezzedex) [inline examples/integration/src/main.rs:226]: Can we handle the other errors? I think the only error that makes sense to panic is `OutOfMemory`.

File: graphics/src/window/compositor.rs [hunk 2]
```diff
@@ -49,5 +52,5 @@ pub trait Compositor: Sized {
         background_color: Color,
         output: &<Self::Renderer as iced_native::Renderer>::Output,
         overlay: &[T],
-    ) -> mouse::Interaction;
+    ) -> Result<mouse::Interaction, ()>;
```
Comment (derezzedex) [inline graphics/src/window/compositor.rs:55]: I think a proper error type here would make sense, maybe just copy `SwapChainError`.

**Review by derezzedex — CHANGES_REQUESTED** [review]
I've left some suggestions, aside from that, looks great. Good job!

File: graphics/src/window/compositor.rs (hunk 2 above)
Comment (BillyDM) [reply graphics/src/window/compositor.rs:55]: Maybe? I think that the `SwapChainError` is only specific to `wgpu` though.

File: examples/integration/src/main.rs (hunk 1 above)
Comment (BillyDM) [reply examples/integration/src/main.rs:226]: Yeah, this makes sense.

File: graphics/src/window/compositor.rs (hunk 2 above)
Comment (BillyDM) [reply graphics/src/window/compositor.rs]: Actually, yeah that could work. I've made those changes.

**Review by hecrj — APPROVED** [review]
Looks great @BillyDM! And great suggestions, @derezzedex.

Thank you both and sorry for the wait! 🙇

**Review by derezzedex — APPROVED** [review]

**Pull request merged**